modules/core/upgrade.py

```diff
@@ -0,0 +1,551 @@
+#!/usr/bin/env python3
+
+import argparse
+from datetime import timedelta
+import math
+import os
+import re
+import shutil
+import subprocess
+import sys
+import time
+
+#
+# Constants
+#
+
+# Tag patterns to filter
+CONSUL_TAG = 'consul'
+NOMAD_SERVER_TAG = 'nomad-server'
+NOMAD_CLIENT_TAG = 'nomad-client'
+VAULT_TAG = 'vault'
+
+# Default addresses
+CONSUL_ADDR = 'http://127.0.0.1:8500'
+NOMAD_ADDR = 'http://127.0.0.1:4646'
+
+# Vault defaults
+VAULT_USERNAME = 'ubuntu'
+VAULT_REMOTE_CA_CERT_PATH = '/etc/ssl/certs/'
+VAULT_LOCAL_ADDR = 'https://127.0.0.1:8200'
+
+SEALED_RE = re.compile(r'Sealed\s+true')
+UNSEALED_RE = re.compile(r'Sealed\s+false')
+
+# Interval of checking the change in set entries
+CHECK_INTERVAL_SECS = 5
+
+# Max time value to allow for ensuring both AWS and services have same set after killing
+TIMEOUT_SECS = 300
+
+# Default number of unseals required
+VAULT_UNSEAL_COUNT = 3
+
+# Commands to ensure to be available before starting
+CMDS = [
+    'aws',
+    'consul',
+    'curl',
+    'nomad',
+    'jq',
+    'scp',
+    'ssh'
+]
+
+#
+# Assertion commands
+#
+
+
+def assert_n_quorum(n):
+    if n == 0:
+        sys.exit(
+            'This process can only be run when there are 3 or more servers to maintain quorum.')
```

Comment:
Hmm I think the naming of the function is weird and misleading (already changed to `assert_kill_count` instead). It's trying to say that you need to have an initial instance count of 3 before you can proceed because quorum requires at least count of 3.

In any case, I have split into `assert_instance_count` to check initial count of 3, and `assert_kill_count` to check that the kill count must be > 0 for clarity.